# NumPy and Pandas - project practice
### eBay dataset, can be found: https://data.world/data-society/used-cars-data 
- contains data regarding used car listings
- 371528 rows × 20 columns


# Aims:
- clean data (the data set I was orignally working on with Jupyter was unclean - however the kaggle version has been cleaned slightly. steps I made to clean are highlighted as comments 
- analyse the used car listing data

Most of all, I just wanted to showcase my progress with both pandas and NumPy. I'm using another scenario provided by Dataquest.io as I am currently following their data scientist pathway. 

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv("../input/autosdataset/autos.csv", encoding="Latin-1")

In [3]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

## General observations
- inconsistent formating for several columns, most notably 'name' with a mixture of full capitals and uncapitalised text
- some values missing for columns, non-null count varied for the 5000 entries
- most common datatype is string (object)
- column format is in camelcase (would prefer snakecase - more standardised)

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

editing column names to improve readability:

In [6]:
autos.columns = ('date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen',)

In [7]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_PS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [8]:
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_PS  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

   odometer  registration_month fuel_type       brand unrepaired_damage  \
0    150000                   0    benzin  volkswagen               NaN   
1    125000                   5    diesel        audi                ja   
2    125000                   8    diesel        jeep               NaN   
3    150000                   6    benzin  volkswagen              nein   
4     90000                   7    diesel       skoda              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

# Data exploration
Initially looking for text columns where all or almost all values are the same (lack of information to analyse). Also looking for any data types that should be converted e.g. numeric data stored as strings that may need cleaning before converting. 

In [9]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_PS  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model       odometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

From the above:
- most sellers are 'privat', with the exception of 1 listing
- most offer types are 'Angebot' (German for 'offer')
- (potentially) unrepaired_damage, most are nein (35232/40171)

datatype issues:
- price, odometer are stored as text when they're actually numeric

In [10]:
autos["price"]

0           480
1         18300
2          9800
3          1500
4          3600
          ...  
371523     2200
371524     1199
371525     9200
371526     3400
371527    28990
Name: price, Length: 371528, dtype: int64

I'll remove the $ and comma:

In [11]:
#the dataset i was originally working on was 'pre-cleaned, in which $ and comma were included' autos["price"] = autos["price"].str.replace("$", "").str.replace(",", "")

change the datatype to int:

In [12]:
autos["price"] = autos["price"].astype(int)

In [13]:
autos.info() #to confirm change to int was successful

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   date_crawled        371528 non-null  object
 1   name                371528 non-null  object
 2   seller              371528 non-null  object
 3   offer_type          371528 non-null  object
 4   price               371528 non-null  int64 
 5   abtest              371528 non-null  object
 6   vehicle_type        333659 non-null  object
 7   registration_year   371528 non-null  int64 
 8   gearbox             351319 non-null  object
 9   power_PS            371528 non-null  int64 
 10  model               351044 non-null  object
 11  odometer            371528 non-null  int64 
 12  registration_month  371528 non-null  int64 
 13  fuel_type           338142 non-null  object
 14  brand               371528 non-null  object
 15  unrepaired_damage   299468 non-null  object
 16  ad

I'll now clean 'odometer' column:

In [14]:
autos["odometer"]

0         150000
1         125000
2         125000
3         150000
4          90000
           ...  
371523     20000
371524    125000
371525    150000
371526    150000
371527     50000
Name: odometer, Length: 371528, dtype: int64

In [15]:
# autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "")

In [16]:
autos["odometer"] = autos["odometer"].astype(int)

In [17]:
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

In [18]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   date_crawled        371528 non-null  object
 1   name                371528 non-null  object
 2   seller              371528 non-null  object
 3   offer_type          371528 non-null  object
 4   price               371528 non-null  int64 
 5   abtest              371528 non-null  object
 6   vehicle_type        333659 non-null  object
 7   registration_year   371528 non-null  int64 
 8   gearbox             351319 non-null  object
 9   power_PS            371528 non-null  int64 
 10  model               351044 non-null  object
 11  odometer_km         371528 non-null  int64 
 12  registration_month  371528 non-null  int64 
 13  fuel_type           338142 non-null  object
 14  brand               371528 non-null  object
 15  unrepaired_damage   299468 non-null  object
 16  ad

In [19]:
autos["odometer_km"]

0         150000
1         125000
2         125000
3         150000
4          90000
           ...  
371523     20000
371524    125000
371525    150000
371526    150000
371527     50000
Name: odometer_km, Length: 371528, dtype: int64

## Analysing price and odometer_km
- determine how many unique values
- min/max values to identify any outliers that might need removal

In [20]:
autos["odometer_km"].unique().shape #to determine how many unique values

(13,)

In [21]:
autos["odometer_km"].describe()

count    371528.000000
mean     125618.688228
std       40112.337051
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [22]:
autos["odometer_km"].value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: odometer_km, dtype: int64

In [23]:
autos["price"].unique().shape

(5597,)

In [24]:
autos["price"].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [25]:
autos["price"].value_counts().sort_index(ascending=True).head(15)
#to view lowest values. chained to .head() because of the amount of unique values

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
Name: price, dtype: int64

In [26]:
autos["price"].value_counts().sort_index(ascending=False).head(15)
#to view highest values

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
Name: price, dtype: int64

removing entries 

In [27]:
autos = autos[autos["price"].between(1,350001)] 

In [28]:
autos["price"].value_counts().sort_index(ascending=False).head(15)

350000    4
349000    1
345000    1
323223    1
300000    1
299000    3
295000    1
294900    1
285000    1
284000    1
279000    1
275000    1
270000    1
265000    1
260000    3
Name: price, dtype: int64

# Date cleaning
- 4 different columns with 'date' information: date_crawled, last_seen, ad_created, registration_year
- currently represented as strings, so need to convert to allow quantitative analysis


In [29]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21

First 10 characters in each column represent the date:

In [30]:
print(autos['date_crawled'].str[:10])

0         2016-03-24
1         2016-03-24
2         2016-03-14
3         2016-03-17
4         2016-03-31
             ...    
371523    2016-03-14
371524    2016-03-05
371525    2016-03-19
371526    2016-03-20
371527    2016-03-07
Name: date_crawled, Length: 360635, dtype: object


To display as percentages instead of counts:
- normalize=True

to rank by date (ascending order):
- chain sort_index()

In [31]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

2016-04-07    0.001617
2016-04-06    0.003128
2016-04-05    0.012780
2016-04-04    0.037628
2016-04-03    0.038812
2016-04-02    0.035094
2016-04-01    0.034145
2016-03-31    0.031872
2016-03-30    0.033535
2016-03-29    0.034126
2016-03-28    0.035063
2016-03-27    0.030227
2016-03-26    0.031974
2016-03-25    0.032800
2016-03-24    0.029914
2016-03-23    0.032002
2016-03-22    0.032493
2016-03-21    0.035682
2016-03-20    0.036400
2016-03-19    0.035271
2016-03-18    0.013119
2016-03-17    0.031647
2016-03-16    0.030205
2016-03-15    0.033424
2016-03-14    0.036330
2016-03-13    0.015783
2016-03-12    0.036242
2016-03-11    0.032773
2016-03-10    0.032645
2016-03-09    0.034115
2016-03-08    0.033469
2016-03-07    0.035657
2016-03-06    0.014483
2016-03-05    0.025547
Name: date_crawled, dtype: float64

In [32]:
autos['date_crawled'].describe

<bound method NDFrame.describe of 0         2016-03-24 11:52:17
1         2016-03-24 10:58:45
2         2016-03-14 12:52:21
3         2016-03-17 16:54:04
4         2016-03-31 17:25:20
                 ...         
371523    2016-03-14 17:48:27
371524    2016-03-05 19:56:21
371525    2016-03-19 18:57:12
371526    2016-03-20 19:41:08
371527    2016-03-07 19:39:19
Name: date_crawled, Length: 360635, dtype: object>

In [33]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

2016-04-07    0.001553
2016-04-06    0.003119
2016-04-05    0.011613
2016-04-04    0.037736
2016-04-03    0.039001
                ...   
2015-08-07    0.000003
2015-06-18    0.000003
2015-06-11    0.000003
2015-03-20    0.000003
2014-03-10    0.000003
Name: ad_created, Length: 114, dtype: float64

In [34]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

2016-04-07    0.130437
2016-04-06    0.218950
2016-04-05    0.126962
2016-04-04    0.025536
2016-04-03    0.025308
2016-04-02    0.024967
2016-04-01    0.023897
2016-03-31    0.024243
2016-03-30    0.023725
2016-03-29    0.023284
2016-03-28    0.022189
2016-03-27    0.016721
2016-03-26    0.015958
2016-03-25    0.019000
2016-03-24    0.019163
2016-03-23    0.018015
2016-03-22    0.020508
2016-03-21    0.020026
2016-03-20    0.019884
2016-03-19    0.016330
2016-03-18    0.006888
2016-03-17    0.028699
2016-03-16    0.016418
2016-03-15    0.016324
2016-03-14    0.012176
2016-03-13    0.008410
2016-03-12    0.023240
2016-03-11    0.012955
2016-03-10    0.011460
2016-03-09    0.009824
2016-03-08    0.007939
2016-03-07    0.005202
2016-03-06    0.004098
2016-03-05    0.001264
Name: last_seen, dtype: float64

In [35]:
autos['registration_year'].describe()

count    360635.000000
mean       2004.433133
std          81.016977
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

### Observations 
- date crawled column suggests the website was crawled between 03/16 to 04/16 
- registration year's maximum is year 9999, and minimum as 1000 (way into the future, and a registration year of 1000 long before cars were invented)
- listing are of 'used' cars, so registration date must be pre-listing date

The ads were created up to 2016, therefore, will restrict data in this column from 1900 - 2016

In [36]:
autos = autos[autos["registration_year"].between(1900,2016)] 

In [37]:
autos["registration_year"].describe()

count    346660.000000
mean       2002.896489
std           7.244719
min        1910.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        2016.000000
Name: registration_year, dtype: float64

# Data cleaning - aggregation
- using loops in Pandas for aggregation 
- i'll use the brand column, and use the values in price, registration year and km
- i'll take the top 10 brands

In [38]:
autos["brand"].value_counts() #to produce series with index labels, though may use series.index

volkswagen        73388
bmw               38088
opel              36888
mercedes_benz     33571
audi              31041
ford              23891
renault           16472
peugeot           10453
fiat               8906
seat               6469
skoda              5438
mazda              5333
smart              4968
citroen            4836
nissan             4714
toyota             4483
hyundai            3457
sonstige_autos     3291
mini               3253
volvo              3171
mitsubishi         2855
honda              2611
kia                2397
suzuki             2206
alfa_romeo         2187
porsche            2153
chevrolet          1741
chrysler           1339
dacia               865
jeep                760
land_rover          751
daihatsu            749
subaru              734
jaguar              601
saab                508
daewoo              505
trabant             488
lancia              451
rover               441
lada                207
Name: brand, dtype: int64

In [39]:
avg_price_by_brand = {}
avg_year_by_brand = {}
avg_km_by_brand = {}

brands = autos["brand"].value_counts().index[:11]

In [40]:
brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda'],
      dtype='object')

In [41]:
for i in brands:
    rows = autos[autos["brand"] == i]
    mean = rows["price"].mean()
    avg_price_by_brand[i] = mean

In [42]:
avg_price_by_brand

{'volkswagen': 5400.188763830599,
 'bmw': 8449.121823146397,
 'opel': 2971.8999132509216,
 'mercedes_benz': 8551.650114682316,
 'audi': 9086.279597951097,
 'ford': 3696.4059687748527,
 'renault': 2437.6054516755707,
 'peugeot': 3267.612647086961,
 'fiat': 2892.9535144846172,
 'seat': 4541.999227083012,
 'skoda': 6530.51084957705}

In [43]:
for i in brands:
    rows = autos[autos["brand"] == i]
    mean_year = rows["registration_year"].mean()
    avg_year_by_brand[i] = mean_year

In [44]:
avg_year_by_brand

{'volkswagen': 2002.4872186188477,
 'bmw': 2003.0101081705525,
 'opel': 2002.1553892864888,
 'mercedes_benz': 2002.0495368025975,
 'audi': 2003.9399503881962,
 'ford': 2002.9697794148424,
 'renault': 2002.8629796017485,
 'peugeot': 2003.8536305366881,
 'fiat': 2002.9440826409163,
 'seat': 2004.734734889473,
 'skoda': 2006.8254873115116}

In [45]:
for i in brands:
    rows = autos[autos["brand"] == i]
    mean_km = rows["odometer_km"].mean()
    avg_km_by_brand[i] = mean_km

In [46]:
avg_km_by_brand

{'volkswagen': 128386.04403989753,
 'bmw': 132800.0945179584,
 'opel': 128722.61982216439,
 'mercedes_benz': 130572.96476125227,
 'audi': 129443.9612125898,
 'ford': 123662.25775396593,
 'renault': 127885.19912578922,
 'peugeot': 124500.62183105329,
 'fiat': 116619.69458791826,
 'seat': 120608.28567011903,
 'skoda': 113593.23280617874}

### Observations
note: I had originally included the top 20 brands, which skewed my results and put 'sontige_auto' branded cars as the most expensive. however, they account for less than 5% of listings, and so I adjusted to include only the top 10 with a greater share of the overall listings.

- audi cars are the most expensive, second to mercendes benz and third to bmw
- cheapest cars are ford and opel
- most of the registered years are in the early 20's, 
- average km appears to be fairly consistent (11,000+ km) 
- just based on the top 10 selection, many are German brands, either reflecting popularity of German cars or (possibly more likely) that the listing were scraped from the German eBay website

To make the process of comparing between aggregates more easier:

In [47]:
brand_series = pd.Series(avg_price_by_brand) #keys in dictionary will become the index in the series 

In [48]:
brand_series

volkswagen       5400.188764
bmw              8449.121823
opel             2971.899913
mercedes_benz    8551.650115
audi             9086.279598
ford             3696.405969
renault          2437.605452
peugeot          3267.612647
fiat             2892.953514
seat             4541.999227
skoda            6530.510850
dtype: float64

In [49]:
brand_series_df = pd.DataFrame(brand_series, columns=['avg_price'])

In [50]:
brand_series_df

avg_price
volkswagen     5400.188764
bmw            8449.121823
opel           2971.899913
mercedes_benz  8551.650115
audi           9086.279598
ford           3696.405969
renault        2437.605452
peugeot        3267.612647
fiat           2892.953514
seat           4541.999227
skoda          6530.510850

now i'll continue with the rest of the dictionaries to:
- convert to series
- assign the series to column in the dataframe above

In [51]:
km_series = pd.Series(avg_km_by_brand)

In [52]:
brand_series_df["avg_km"] = km_series

In [53]:
year_series = pd.Series(avg_year_by_brand)

In [54]:
brand_series_df["avg_year"] = year_series

In [55]:
brand_series_df

avg_price         avg_km     avg_year
volkswagen     5400.188764  128386.044040  2002.487219
bmw            8449.121823  132800.094518  2003.010108
opel           2971.899913  128722.619822  2002.155389
mercedes_benz  8551.650115  130572.964761  2002.049537
audi           9086.279598  129443.961213  2003.939950
ford           3696.405969  123662.257754  2002.969779
renault        2437.605452  127885.199126  2002.862980
peugeot        3267.612647  124500.621831  2003.853631
fiat           2892.953514  116619.694588  2002.944083
seat           4541.999227  120608.285670  2004.734735
skoda          6530.510850  113593.232806  2006.825487

In [56]:
brand_series_df.sort_values(by='avg_price', axis=0, ascending=False)

avg_price         avg_km     avg_year
audi           9086.279598  129443.961213  2003.939950
mercedes_benz  8551.650115  130572.964761  2002.049537
bmw            8449.121823  132800.094518  2003.010108
skoda          6530.510850  113593.232806  2006.825487
volkswagen     5400.188764  128386.044040  2002.487219
seat           4541.999227  120608.285670  2004.734735
ford           3696.405969  123662.257754  2002.969779
peugeot        3267.612647  124500.621831  2003.853631
opel           2971.899913  128722.619822  2002.155389
fiat           2892.953514  116619.694588  2002.944083
renault        2437.605452  127885.199126  2002.862980

Isn't any significant difference between mileage and price - slight increase in average km with more expensive models. 